File: apps/demo-fs-ts-nextjs/test/unit/be-steps/registration.steps.ts
import path from "path";
import { loadFeature, describeFeature } from "@amiceli/vitest-cucumber";
import { expect } from "vitest";
import { createTestContext, registerUser, type TestContext } from "./helpers/test-context";

const feature = await loadFeature(
  path.resolve(__dirname, "../../../../../specs/apps/demo/be/gherkin/user-lifecycle/registration.feature"),
);

describeFeature(feature, ({ Scenario, Background }) => {
  let ctx: TestContext;

  Background(({ Given }) => {
    Given("the API is running", () => {
      ctx = createTestContext();
    });
  });

  Scenario("Successful registration returns created user profile without password", ({ When, Then, And }) => {
    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "alice@example.com", "password": "Str0ng#Pass1" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "alice@example.com", password: "Str0ng#Pass1" },
          null,
        );
      },
    );

    Then("the response status code should be 201", () => {
      expect(ctx.response!.status).toBe(201);
    });

    And('the response body should contain "username" equal to "alice"', () => {
      expect((ctx.response!.body as Record<string, unknown>).username).toBe("alice");
    });

    And('the response body should not contain a "password" field', () => {
      expect(ctx.response!.body).not.toHaveProperty("password");
    });
  });

  Scenario("Successful registration response includes non-null user ID", ({ When, Then, And }) => {
    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "alice@example.com", "password": "Str0ng#Pass1" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "alice@example.com", password: "Str0ng#Pass1" },
          null,
        );
      },
    );

    Then("the response status code should be 201", () => {
      expect(ctx.response!.status).toBe(201);
    });

    And('the response body should contain a non-null "id" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).id).toBeDefined();
    });
  });

  Scenario("Reject registration when username already exists", ({ Given, When, Then, And }) => {
    Given('a user "alice" is registered with email "alice@example.com" and password "Str0ng#Pass1"', async () => {
      await registerUser(ctx, "alice", "alice@example.com", "Str0ng#Pass1");
    });

    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "new@example.com", "password": "Str0ng#Pass1" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "new@example.com", password: "Str0ng#Pass1" },
          null,
        );
      },
    );

    Then("the response status code should be 409", () => {
      expect(ctx.response!.status).toBe(409);
    });

    And("the response body should contain an error message about duplicate username", () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("username");
    });
  });

  Scenario("Reject registration with invalid email format", ({ When, Then, And }) => {
    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "not-an-email", "password": "Str0ng#Pass1" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "not-an-email", password: "Str0ng#Pass1" },
          null,
        );
      },
    );

    Then("the response status code should be 400", () => {
      expect(ctx.response!.status).toBe(400);
    });

    And('the response body should contain a validation error for "email"', () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("email");
    });
  });

  Scenario("Reject registration with empty password", ({ When, Then, And }) => {
    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "alice@example.com", "password": "" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "alice@example.com", password: "" },
          null,
        );
      },
    );

    Then("the response status code should be 400", () => {
      expect(ctx.response!.status).toBe(400);
    });

    And('the response body should contain a validation error for "password"', () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("password");
    });
  });

  Scenario("Reject registration with weak password — no uppercase letter", ({ When, Then, And }) => {
    When(
      'the client sends POST /api/v1/auth/register with body { "username": "alice", "email": "alice@example.com", "password": "str0ng#pass1" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/auth/register",
          { username: "alice", email: "alice@example.com", password: "str0ng#pass1" },
          null,
        );
      },
    );

    Then("the response status code should be 400", () => {
      expect(ctx.response!.status).toBe(400);
    });

    And('the response body should contain a validation error for "password"', () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("password");
    });
  });
});

File: apps/crud-be-elixir-phoenix/test/integration/steps/registration_steps.exs
defmodule CrudBeExphWeb.Integration.RegistrationSteps do
  use Cabbage.Feature, async: false, file: "user-lifecycle/registration.feature"

  use CrudBeExph.DataCaseIntegration

  alias CrudBeExph.Integration.Helpers
  alias CrudBeExph.Integration.ServiceLayer

  @moduletag :integration

  defgiven ~r/^the API is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered with email "(?<email>[^"]+)" and password "(?<password>[^"]+)"$/,
           %{username: username, email: email, password: password},
           state do
    Helpers.register_user!(username, email, password)
    {:ok, state}
  end

  defwhen ~r/^the client sends POST .api.v1.auth.register with body \{ "username": "(?<username>[^"]+)", "email": "(?<email>[^"]+)", "password": "(?<password>[^"]*)" \}$/,
          %{username: username, email: email, password: password},
          state do
    response =
      ServiceLayer.register(%{"username" => username, "email" => email, "password" => password})

    {:ok, Map.put(state, :response, response)}
  end

  defthen ~r/^the response status code should be (?<code>\d+)$/,
          %{code: code},
          %{response: response} = state do
    assert response.status == String.to_integer(code)
    {:ok, state}
  end

  defthen ~r/^the response body should contain "(?<field>[^"]+)" equal to "(?<value>[^"]+)"$/,
          %{field: field, value: value},
          %{response: response} = state do
    assert response.body[field] == value
    {:ok, state}
  end

  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Successful registration returns created user profile without password
  defthen ~r/^the response body should not contain a "(?<field>[^"]+)" field$/,
          %{field: field},
          %{response: response} = state do
    refute Map.has_key?(response.body, field)
    {:ok, state}
  end

  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Successful registration response includes non-null user ID
  defthen ~r/^the response body should contain a non-null "(?<field>[^"]+)" field$/,
          %{field: field},
          %{response: response} = state do
    assert Map.has_key?(response.body, field)
    assert response.body[field] != nil
    {:ok, state}
  end

  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Reject registration when username already exists
  defthen ~r/^the response body should contain an error message about duplicate username$/,
          _vars,
          %{response: response} = state do
    assert response.body["message"] =~ ~r/[Uu]sername.*exist|already|[Dd]uplicate/i
    {:ok, state}
  end

  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Reject registration with invalid email format
  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Reject registration with empty password
  # @covers specs/apps/crud/behavior/crud-be/gherkin/user-lifecycle/registration.feature:Reject registration with weak password — no uppercase letter
  defthen ~r/^the response body should contain a validation error for "(?<field>[^"]+)"$/,
          %{field: field},
          %{response: response} = state do
    assert Map.has_key?(response.body, "errors")
    errors = response.body["errors"]
    assert Map.has_key?(errors, field)
    assert errors[field] != []
    {:ok, state}
  end
end
